### PR TITLE
[FIX] project_timesheet_holidays: changing public holidays deletes auto-cancelled linked timesheets

### PR DESCRIPTION
**Pull request by chrisduvivier**
Description of the issue/feature this PR addresses:
-------------------

Auto-cancelling Time Offs when editing public holidays was added in Jun 2022 ([see PR](https://github.com/odoo/odoo/pull/81225)) , but the cancellation of Time Offs wasn't updated to also remove the link to timesheets. If the user attempts to manually delete a timesheet linked to a Time Off (`holiday_id`​ in `hr.leave`​), they will be asked to cancel the Time Off which by this point had already been auto-cancelled, and as long as the link exists (regardless of cancellation) the timesheet cannot be deleted from the user frontend.

This PR adds an override call to unlink the timesheets before the leave (the inital validated timeoff) is force canceled.

Current behavior before PR:
-------------------
Steps to reproduce:
1. Create and validate an arbitrary Time Off, checking that it also creates a timesheet entry.
2. Create a new public holiday spanning the entire duration of the Time Off created in (1)
3. Attempt to delete the timesheet entry created in (1), if it is not already deleted.

Although you will be asked to cancel the timeoff first, that is not possible because the timeoff will be auto-cancelled by this point, making the deletion of the timesheet impossible.

Desired behavior after PR is merged:
-------------------

At step 3, user should be able to delete the timesheet associated to the timeoff.

Task-3697074

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thanks for your work :zap:

**Comment by chrisduvivier** [comment]
@auon-odoo Thank you for the review! I applied the changes and pushed.

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 1]
```diff
@@ -139,3 +139,12 @@ def _action_user_cancel(self, reason):
...
         self._check_missing_global_leave_timesheets()
         return res
+
+    def _force_cancel(self, *args, **kwargs):
+        """
+        override this method to reevaluate timesheets after the leaves are updated via force cancel
+        """
+        super()._force_cancel(*args, **kwargs)
+        timesheets = self.sudo().timesheet_ids
+        timesheets.write({'holiday_id': False})
```
Comment (auon-odoo) [inline addons/project_timesheet_holidays/models/hr_holidays.py:149]:
```suggestion
        timesheets.holiday_id = False
```
looks better, but it is a detail

**Review by auon-odoo — APPROVED** [review]
Lgtm, thanks :+1:

**Comment by chrisduvivier** [comment]
Thanks for the suggestion, I changed it and rebased the branch @auon-odoo

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 2]
```diff
@@ -135,7 +135,16 @@ def action_refuse(self):
     def _action_user_cancel(self, reason):
         res = super()._action_user_cancel(reason)
         timesheets = self.sudo().timesheet_ids
-        timesheets.write({'holiday_id': False})
+        timesheets.holiday_id = False
```
Comment (MissingNoShiny) [inline addons/project_timesheet_holidays/models/hr_holidays.py:138]: It's better to avoid unnecessary diff in stable

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 3]
```diff
@@ -135,7 +135,16 @@ def action_refuse(self):
...
         self._check_missing_global_leave_timesheets()
         return res
+
+    def _force_cancel(self, *args, **kwargs):
+        """
+        override this method to reevaluate timesheets after the leaves are updated via force cancel
+        """
+        super()._force_cancel(*args, **kwargs)
+        timesheets = self.sudo().timesheet_ids
+        timesheets.holiday_id = False
```
Comment (MissingNoShiny) [inline addons/project_timesheet_holidays/models/hr_holidays.py:149]: Is this line useful, since we unlink the timesheets right after? I think the default delete policy should be “set null” in this case

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 4]
```diff
@@ -135,7 +135,16 @@ def action_refuse(self):
...
-        timesheets.write({'holiday_id': False})
+        timesheets.holiday_id = False
         timesheets.unlink()
         self._check_missing_global_leave_timesheets()
         return res
+
+    def _force_cancel(self, *args, **kwargs):
+        """
+        override this method to reevaluate timesheets after the leaves are updated via force cancel
+        """
```
Comment (MissingNoShiny) [inline addons/project_timesheet_holidays/models/hr_holidays.py:146]: This looks like a docstring, I would add a simple # comment after the call to super

**Review by MissingNoShiny — CHANGES_REQUESTED** [review]
Thanks for your work, I left a few comments :slightly_smiling_face:

File: addons/project_timesheet_holidays/models/hr_holidays.py (hunk 2 above)
Comment (chrisduvivier) [reply addons/project_timesheet_holidays/models/hr_holidays.py:138]: Thank you for the review Vincent, I will revert the change.

File: addons/project_timesheet_holidays/models/hr_holidays.py (hunk 3 above)
Comment (chrisduvivier) [reply addons/project_timesheet_holidays/models/hr_holidays.py:149]: This line was actually necessary for this task, and removing this line raises an error: `You cannot delete timesheets linked to time off. Please, cancel the time off instead`

**Review by MissingNoShiny — APPROVED** [review]
LGTM, thanks!
robodoo r+
